From 27b485a276c97cbd3d60ec2a05deb8af2eee54a3 Mon Sep 17 00:00:00 2001
From: Szabolcs Nagy <nsz@port70.net>
Date: Wed, 26 Aug 2026 08:18:57 +0200
Subject: math: fix expl on x86

expl asm special cased |x|>=16384 and used 2^trunc(x) then,
but this was wrong for x<=-16384 when 2^trunc(x) doesn't
underflow to 0. fixed by bumping the threshold up to 32768.

Reported-by: Paul Zimmermann <Paul.Zimmermann@inria.fr>
---
 src/math/i386/expl.s   | 10 +++++-----
 src/math/x32/expl.s    | 10 +++++-----
 src/math/x86_64/expl.s | 10 +++++-----
 3 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/src/math/i386/expl.s b/src/math/i386/expl.s
index b5124e8f..1cd20387 100644
--- a/src/math/i386/expl.s
+++ b/src/math/i386/expl.s
@@ -8,17 +8,17 @@
 expl:
 	fldt 4(%esp)
 
-		# interesting case: 0x1p-32 <= |x| < 16384
-		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+13]
+		# interesting case: 0x1p-32 <= |x| < 0x1p15
+		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+14]
 	mov 12(%esp), %ax
 	or $0x8000, %ax
-	sub $0xbfdf, %ax
-	cmp $45, %ax
+	sub $0xbfe0, %ax
+	cmp $46, %ax
 	jbe 2f
 	test %ax, %ax
 	fld1
 	js 1f
-		# if |x|>=0x1p14 or nan return 2^trunc(x)
+		# if |x|>=0x1p15 or nan return 2^trunc(x)
 	fscale
 	fstp %st(1)
 	ret
diff --git a/src/math/x32/expl.s b/src/math/x32/expl.s
index 369f7bd2..b5f3ded0 100644
--- a/src/math/x32/expl.s
+++ b/src/math/x32/expl.s
@@ -8,17 +8,17 @@
 expl:
 	fldt 8(%esp)
 
-		# interesting case: 0x1p-32 <= |x| < 16384
-		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+13]
+		# interesting case: 0x1p-32 <= |x| < 0x1p15
+		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+14]
 	mov 16(%esp), %ax
 	or $0x8000, %ax
-	sub $0xbfdf, %ax
-	cmp $45, %ax
+	sub $0xbfe0, %ax
+	cmp $46, %ax
 	jbe 2f
 	test %ax, %ax
 	fld1
 	js 1f
-		# if |x|>=0x1p14 or nan return 2^trunc(x)
+		# if |x|>=0x1p15 or nan return 2^trunc(x)
 	fscale
 	fstp %st(1)
 	ret
diff --git a/src/math/x86_64/expl.s b/src/math/x86_64/expl.s
index 798261d2..b5a0bac9 100644
--- a/src/math/x86_64/expl.s
+++ b/src/math/x86_64/expl.s
@@ -8,17 +8,17 @@
 expl:
 	fldt 8(%rsp)
 
-		# interesting case: 0x1p-32 <= |x| < 16384
-		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+13]
+		# interesting case: 0x1p-32 <= |x| < 0x1p15
+		# check if (exponent|0x8000) is in [0xbfff-32, 0xbfff+14]
 	mov 16(%rsp), %ax
 	or $0x8000, %ax
-	sub $0xbfdf, %ax
-	cmp $45, %ax
+	sub $0xbfe0, %ax
+	cmp $46, %ax
 	jbe 2f
 	test %ax, %ax
 	fld1
 	js 1f
-		# if |x|>=0x1p14 or nan return 2^trunc(x)
+		# if |x|>=0x1p15 or nan return 2^trunc(x)
 	fscale
 	fstp %st(1)
 	ret
-- 
cgit v1.2.1

